File: app/src/main/java/io/github/ovso/righttoknow/ui/main/video/VideoFragment.java
package io.github.ovso.righttoknow.ui.main.video;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;
import io.github.ovso.righttoknow.R;
import io.github.ovso.righttoknow.data.network.model.video.SearchItem;
import io.github.ovso.righttoknow.framework.DaggerFragment;
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView;
import io.github.ovso.righttoknow.ui.base.OnEndlessRecyclerScrollListener;
import io.github.ovso.righttoknow.ui.base.OnRecyclerViewItemClickListener;
import io.github.ovso.righttoknow.ui.base.VideoRecyclerView;
import io.github.ovso.righttoknow.ui.video.LandscapeVideoActivity;
import io.github.ovso.righttoknow.ui.video.PortraitVideoActivity;
import javax.inject.Inject;

public class VideoFragment extends DaggerFragment implements VideoFragmentPresenter.View,
    OnEndlessRecyclerScrollListener.OnLoadMoreListener,
    OnRecyclerViewItemClickListener<SearchItem> {
  @BindView(R.id.swipe_refresh) SwipeRefreshLayout swipeRefresh;
  @BindView(R.id.recyclerview) VideoRecyclerView recyclerView;
  @Inject VideoFragmentPresenter presenter;
  private BaseAdapterView adapterView;

  public static VideoFragment newInstance() {
    VideoFragment f = new VideoFragment();
    return f;
  }

  @Override public void onActivityCreated(@Nullable Bundle savedInstanceState) {
    super.onActivityCreated(savedInstanceState);
    setHasOptionsMenu(true);
    presenter.onActivityCreated(savedInstanceState);
  }

  @Override public int getLayoutResId() {
    return R.layout.fragment_video;
  }

  @Override public void setRecyclerView() {
    LinearLayoutManager layout = new LinearLayoutManager(getContext());
    VideoAdapter adapter = new VideoAdapter();
    presenter.setAdapterDataModel(adapter);
    adapterView = adapter;
    recyclerView.setLayoutManager(layout);
    recyclerView.setAdapter(adapter);
    recyclerView.addOnScrollListener(
        new OnEndlessRecyclerScrollListener
            .Builder((LinearLayoutManager) recyclerView.getLayoutManager(), this)
            .setVisibleThreshold(20)
            .build()
    );
    recyclerView.setOnItemClickListener(this);
  }

  @Override public void refresh() {
    adapterView.refresh();
  }

  @Override public void setRefreshLayout() {
    swipeRefresh.setOnRefreshListener(() -> presenter.onRefresh());
    swipeRefresh.setColorSchemeResources(R.color.colorPrimary);
  }

  @Override public void showLoading() {
    swipeRefresh.setRefreshing(true);
  }

  @Override public void hideLoading() {
    swipeRefresh.setRefreshing(false);
  }

  @Override public void showMessage(int resId) {
    Toast.makeText(getContext(), resId, Toast.LENGTH_SHORT).show();
  }

  @Override public void setLoaded() {
    recyclerView.getOnEndlessRecyclerScrollListener().setLoaded();
  }

  @Override public void onDestroyView() {
    presenter.onDestroyView();
    super.onDestroyView();
  }

  @Override public void onResume() {
    super.onResume();
    getActivity().setTitle(R.string.title_video);
  }

  @Override public void onActivityResult(int requestCode, int resultCode, Intent data) {
    super.onActivityResult(requestCode, resultCode, data);
    presenter.onActivityResult(requestCode, resultCode, data);
  }

  @Override public void onLoadMore() {
    presenter.onLoadMore();
  }

  @Override public void onItemClick(View view, SearchItem data, int itemPosition) {
    presenter.onItemClick(data);
  }

  @Override public void showVideoTypeDialog(DialogInterface.OnClickListener onClickListener) {
    new android.app.AlertDialog.Builder(getContext()).setMessage(
        R.string.please_select_the_player_mode)
        .setPositiveButton(R.string.portrait_mode,
            onClickListener)
        .setNeutralButton(R.string.landscape_mode, onClickListener)
        .setNegativeButton(android.R.string.cancel, onClickListener)
        .show();
  }

  @Override public void showPortraitVideo(String videoId) {
    Intent intent = new Intent(getContext(), PortraitVideoActivity.class);
    intent.putExtra("video_id", videoId);
    startActivity(intent);
  }

  @Override public void showLandscapeVideo(String videoId) {
    Intent intent = new Intent(getContext(), LandscapeVideoActivity.class);
    intent.putExtra("video_id", videoId);
    startActivity(intent);
  }

  @Override public void showYoutubeUseWarningDialog() {
    new android.app.AlertDialog.Builder(getActivity()).setMessage(R.string.youtube_use_warning)
        .setPositiveButton(android.R.string.ok, null)
        .show();
  }

  @Override public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
    menu.findItem(R.id.option_menu_sort).setVisible(false);
    super.onCreateOptionsMenu(menu, inflater);
  }

}

File: app/src/main/java/io/github/ovso/righttoknow/ui/main/video/VideoFragment.kt
package io.github.ovso.righttoknow.ui.main.video

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import io.github.ovso.righttoknow.R
import io.github.ovso.righttoknow.data.network.VideoRequest
import io.github.ovso.righttoknow.data.network.model.video.SearchItem
import io.github.ovso.righttoknow.exts.launchActivity
import io.github.ovso.righttoknow.framework.BaseFragment
import io.github.ovso.righttoknow.framework.ad.MyAdView
import io.github.ovso.righttoknow.framework.adapter.BaseAdapterView
import io.github.ovso.righttoknow.ui.base.OnEndlessRecyclerScrollListener
import io.github.ovso.righttoknow.ui.base.OnEndlessRecyclerScrollListener.OnLoadMoreListener
import io.github.ovso.righttoknow.ui.base.OnRecyclerViewItemClickListener
import io.github.ovso.righttoknow.ui.video.LandscapeVideoActivity
import io.github.ovso.righttoknow.ui.video.PortraitVideoActivity
import io.github.ovso.righttoknow.utils.ResourceProvider
import io.github.ovso.righttoknow.utils.SchedulersFacade
import kotlinx.android.synthetic.main.fragment_video.*

class VideoFragment : BaseFragment(), VideoFragmentPresenter.View, OnLoadMoreListener,
  OnRecyclerViewItemClickListener<SearchItem> {
  private lateinit var presenter: VideoFragmentPresenter
  private var adapterView: BaseAdapterView? = null
  override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)
    setHasOptionsMenu(true)
    presenter = VideoFragmentPresenterImpl(
      this,
      MyAdView.getInterstitalAd(requireContext()),
      VideoRequest(),
      ResourceProvider(requireContext()),
      SchedulersFacade
    )

    presenter.onActivityCreated()
  }

  override fun getLayoutResId(): Int {
    return R.layout.fragment_video
  }

  override fun setRecyclerView() {
    val layout = LinearLayoutManager(context)
    val adapter = VideoAdapter()
    presenter.setAdapterDataModel(adapter)
    adapterView = adapter
    rv_video.layoutManager = layout
    rv_video.adapter = adapter
    rv_video.addOnScrollListener(
      OnEndlessRecyclerScrollListener.Builder(rv_video.layoutManager as LinearLayoutManager, this)
        .setVisibleThreshold(20)
        .build()
    )
    rv_video.setOnItemClickListener(this)
  }

  override fun refresh() {
    adapterView!!.refresh()
  }

  override fun setRefreshLayout() {
    srl_video.setOnRefreshListener { presenter.onRefresh() }
    srl_video.setColorSchemeResources(R.color.colorPrimary)
  }

  override fun navigateToPlayer(videoId: String) {
    launchActivity<PortraitVideoActivity> {
      putExtra("video_id", videoId)
    }
  }

  override fun showLoading() {
    srl_video.isRefreshing = true
  }

  override fun hideLoading() {
    srl_video.isRefreshing = false
  }

  override fun showMessage(resId: Int) {
    Toast.makeText(context, resId, Toast.LENGTH_SHORT).show()
  }

  override fun setLoaded() {
    rv_video.onEndlessRecyclerScrollListener?.setLoaded()
  }

  override fun onDestroyView() {
    presenter.onDestroyView()
    super.onDestroyView()
  }

  override fun onResume() {
    super.onResume()
    activity!!.setTitle(R.string.title_video)
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    presenter.onActivityResult(requestCode, resultCode, data)
  }

  override fun onLoadMore() {
    presenter.onLoadMore()
  }

  override fun onItemClick(view: View, data: SearchItem, itemPosition: Int) {
    presenter.onItemClick(data)
  }

  override fun showVideoTypeDialog(onClickListener: DialogInterface.OnClickListener) {
    AlertDialog.Builder(context).setMessage(
      R.string.please_select_the_player_mode
    )
      .setPositiveButton(
        R.string.portrait_mode,
        onClickListener
      )
      .setNeutralButton(R.string.landscape_mode, onClickListener)
      .setNegativeButton(android.R.string.cancel, onClickListener)
      .show()
  }

  override fun showPortraitVideo(videoId: String) {
    val intent = Intent(context, PortraitVideoActivity::class.java)
    intent.putExtra("video_id", videoId)
    startActivity(intent)
  }

  override fun showLandscapeVideo(videoId: String) {
    val intent = Intent(context, LandscapeVideoActivity::class.java)
    intent.putExtra("video_id", videoId)
    startActivity(intent)
  }

  override fun showYoutubeUseWarningDialog() {
    AlertDialog.Builder(activity).setMessage(R.string.youtube_use_warning)
      .setPositiveButton(android.R.string.ok, null)
      .show()
  }

  override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
    menu.findItem(R.id.option_menu_sort).isVisible = false
    super.onCreateOptionsMenu(menu, inflater)
  }

  companion object {
    fun newInstance(): VideoFragment {
      return VideoFragment()
    }
  }
}
